### PR TITLE
chore: Remove global externalURL watcher

### PR DESCRIPTION
**Pull request by eseliger**
This removes the need to register a global URL watcher. Instead, the conf package now returns a cached version, and also makes a copy of it so it's impossible to accidentally modify it.

This makes it safe in non-cmd/frontend packages to use this.

Test plan: All tests are still passing.

### CODE REVIEW

**Comment by eseliger** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64058?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#64058** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64058?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#64056** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64056?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#64055** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64055?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#64054** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64054?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#63655** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63655?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#63651** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63651?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#63650** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63650?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#63649** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63649?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#63648** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63648?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#64012** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64012?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#63743** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63743?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @eseliger and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by camdencheek — APPROVED** [review]
LGTM! I'm a tiny bit concerned about introducing synchronization given how frequently it looks like this is called, but probably not worth optimizing in advance

**Comment by eseliger** [comment]
> I'm a tiny bit concerned about introducing synchronization given how frequently it looks like this is called, but probably not worth optimizing in advance

Hmm yea good point, let's watch the contention graph in GCP Profiler and we can switch those that are too frequent so they're locking up to parse it themselves(?)

**Comment by camdencheek** [comment]
> we can switch those that are too frequent

We could probably also just use an atomic pointer to swap in a new value, then the parsing and copying all happens without any mutex.

File: internal/conf/computed.go [hunk 1]
```diff
@@ -434,6 +437,62 @@ func ProductResearchPageEnabled() bool {
...
+
+var (
+	lastParsedURL *atomic.Pointer[lastParsedURLValue] = atomic.NewPointer(&lastParsedURLValue{
+		url: &url.URL{
+			Scheme: "http",
+			Host:   "example.com",
+		},
+		confValue: "",
+	})
+)
```
Comment (camdencheek) [inline internal/conf/computed.go:453]: nit: no need for the block. Shoudl just be able to do `var lastParsedURL = atomic.NewPointer(...)`

**Comment by eseliger** [comment]
Switched to atomic.Pointer to prevent locking, thanks for the suggestion!

File: internal/conf/computed.go [hunk 2]
```diff
@@ -434,6 +437,62 @@ func ProductResearchPageEnabled() bool {
...
+		parsed, err := url.Parse(urlString)
+		if err != nil {
+			log.Scoped("conf.ExternalURL").Error("failed to parse external URL", log.Error(err), log.String("externalURL", urlString))
+			return cloneURL(lastParsed.url)
+		}
+		lastParsed = &lastParsedURLValue{
+			url:       parsed,
+			confValue: urlString,
+		}
+		lastParsedURL.Store(lastParsed)
```
Comment (camdencheek) [inline internal/conf/computed.go:476]: nit: this looks correct to me, but can we just inline this so I don't have to convince myself we're not writing into the `lastParsed` pointer?

```suggestion
		lastParsedURL.store(&lastParsedURLValue{
			url:       parsed,
			confValue: urlString,
		})		
```

**Review by camdencheek — APPROVED** [review]

File: internal/conf/computed.go (hunk 2 above)
Comment (eseliger) [reply internal/conf/computed.go:476]: yes, just need to also return the new parsed value then otherwise it'll return the stale data below 😬

Comment (camdencheek) [reply internal/conf/computed.go:476]: Aah, gotcha. Missed why that exists. Ignore me then

**Pull request merged**